### PR TITLE
Debugging remote

### PR DESCRIPTION
**Pull request by franciscolopezsancho**
### What changes were proposed in this pull request?
adding remote debug capability 

### Why are the changes needed?
UX

### Does this PR introduce any user-facing change?
yes, the user can attach a debugger

### How was this patch tested?
adding the pertinent configuration and adding breaking points into IntelliJ. Done for all the base images. Breaking points behaved as usual.

### CODE REVIEW

File: docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+Once this is added and the application is deployed we'll need to do two things. 
+
+
+1. Adding a remote configuration in the IDE of our choice. When adding this remote Run/Debug configuration we have to make sure that it has these arguments for the remote JVM.
+	
+	-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005
+
+ Also, this remote configuration shall point to 'localhost' in the port of our choice. 
+
+2. Kubernetes port forwarding from our local to the selected streamlet. Let's say the port of our choice in the IDE is 5005 and our application is called 'example-application'. Then we will need to port forwarding like the following:
```
Comment (andreaTP) [inline docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc:39]: Does this step work even if we don't expose the port while deploying the streamlet? Is it already exposed by default by the operator?

File: docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+include::ROOT:partial$include.adoc[]
+
+
+Debugging a remote application is possible just by applying the https://docs.oracle.com/javase/8/docs/technotes/guides/jpda/conninv.html#Invocation[standard] configuration to debugging a remote JVM. That is, adding this conf to JAVA_OPTS through our xref:cloudflow-configuration.adoc[configuration model]. 
+	
+	-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005
+
+
+NOTE: the address can NOT be changed, it shall remain 5005.
```
Comment (andreaTP) [inline docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc:11]: we can bind the address to an sbt `settingKey`, if a user wants to debug multiple applications or in different environments, this sounds like a limitation, WDYT?

File: core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowAkkaPlugin.scala [hunk 3]
```diff
@@ -90,7 +90,8 @@ object CloudflowAkkaPlugin extends AutoPlugin {
         Instructions.User(UserInImage),
         Instructions.EntryPoint.exec(Seq("bash", "/opt/akka-entrypoint.sh")),
         Instructions.Copy(sources = Seq(CopyFile(depJarsDir)), destination = OptAppDir, chown = Some(userAsOwner(UserInImage))),
-        Instructions.Copy(sources = Seq(CopyFile(appJarsDir)), destination = OptAppDir, chown = Some(userAsOwner(UserInImage)))
+        Instructions.Copy(sources = Seq(CopyFile(appJarsDir)), destination = OptAppDir, chown = Some(userAsOwner(UserInImage))),
+        Instructions.Expose(Seq(5005))
```
Comment (andreaTP) [inline core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowAkkaPlugin.scala:94]: Thinking twice about it, I think we should make this feature optional.
For security reasons, I can easily see customers that do want to not expose ports in production environments.

**Review by andreaTP — COMMENTED** [review]
I left a few suggestions, also, we can easily build up a simple `scripted` test to at least verify that the port is exposed at the docker level to avoid regressions in the future.
Available to help in case!

**Comment by RayRoestenburg** [comment]
Fails on scalafmt.

File: core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowAkkaPlugin.scala (hunk 3 above)
Comment (RayRoestenburg) [reply core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowAkkaPlugin.scala:94]: Can we put this in as an optional sbt setting?

**Review by RayRoestenburg — COMMENTED** [review]
Would be nice to have the debug port as a setting in sbt.

File: docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+----
+
+Once this is added and the application is deployed we'll need to do two things. 
+
+
+1. Adding a remote configuration in the IDE of our choice. When adding this remote Run/Debug configuration we have to make sure that it has these arguments for the remote JVM.
+	
+	-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=*:5005
+
+ Also, this remote configuration shall point to 'localhost' in the port of our choice.
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc:37]:
```suggestion
The above remote configuration example points to 'localhost' and port 5005. 
```
Is maybe better?

File: docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            { name = "JAVA_OPTS" 
+              value = "-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005"
+            }
+          ]
+    }
+  }
+}
+----
+
+Once this is added and the application is deployed we'll need to do two things.
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc:30]:
```suggestion
Once this is added and the application is deployed we'll need to do two things:
```

File: docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+          ]
+    }
+  }
+}
+----
+
+Once this is added and the application is deployed we'll need to do two things. 
+
+
+1. Adding a remote configuration in the IDE of our choice. When adding this remote Run/Debug configuration we have to make sure that it has these arguments for the remote JVM.
```
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc:33]:
```suggestion
1. Add a remote configuration in the IDE of our choice. When adding this remote Run/Debug configuration we have to make sure that it has these arguments for the remote JVM.
```

File: docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc (hunk 1 above)
Comment (RayRoestenburg) [inline docs/shared-content-source/docs/modules/develop/pages/debugging-remote.adoc:39]:
```suggestion
2. Port-forward from our local to the selected streamlet in the Kubernetes cluster. Let's say the port of our choice in the IDE is 5005 and our application is called 'example-application'. Then we will need to port-forward like in the following example:
```

**Comment by franciscolopezsancho** [comment]
I'll pursue open port and passing JVM params all through configuration at deploy time.

**Pull request closed without merging**